torch/utils/data/datapipes/iter/combining.py

```diff
@@ -510,9 +511,18 @@ def __init__(self, *datapipes: IterDataPipe):
         self.length = None
 
     def __iter__(self) -> Iterator[Tuple[T_co]]:
-        for data in zip(*self.datapipes):
+        iterators = [iter(datapipe) for datapipe in self.datapipes]
+        for data in zip(*iterators):
             yield data
 
+        unused = []
+        for iterator in iterators:
+            unused += list(iterator)
+
+        # TODO(VitalyFedyunin): This should be Exception or warning when torchdata.debug is enabled
+        for item in unused:
+            StreamWrapper.close_streams(item)
```

Comment:
I am not sure about the behavior of generator functions. When an not fully consumed iterator is deconstructed, I think the iterator would be destroyed. Technically, these `StreamWrapper` in `unused` variable should not even be created.
Do we need to handle this case?

Reply:
I think there are two options here:
1. Do `try...finally` within `__iter__`, where the `finally` clause does the clean up.
  - This should work regardless how/why the deconstruction of the iterator happens. 
I just tried this with `IterableWrapper`.
```python
    def __iter__(self):
        try:
            source_data = self.iterable
            if self.deepcopy:
                try:
                    source_data = copy.deepcopy(self.iterable)
                # For the case that data cannot be deep-copied,
                # all in-place operations will affect iterable variable.
                # When this DataPipe is iterated second time, it will
                # yield modified items.
                except TypeError:
                    warnings.warn(
                        "The input iterable can not be deepcopied, "
                        "please be aware of in-place modification would affect source data."
                    )
            for data in source_data:
                yield data
        finally:
            print("FINALLLY")
```
```python
dp = IterableWrapper(range(10))
it = iter(dp)
print(next(it))
print("About to create a new iterator")
it = iter(dp)
print("Created a new iterator")
```
```
0
About to create a new iterator
FINALLLY
Created a new iterator
```
2. We can put the clean up logic with `hook_iterator` or `reset` method of `IterDataPipe` whenever that is triggered.

Reply:
I like the idea of `finally:` it will produce readable code and can be placed as a development pattern

Reply:
Done